### PR TITLE
Group primitives and improve tests and API

### PR DESCRIPTION
**Pull request by Rigidity**

### CODE REVIEW

File: crates/chia-sdk-driver/src/primitives/cat.rs [hunk 1]
```diff
@@ -462,6 +459,55 @@ mod tests {
...
     }
 
+    #[test]
+    fn test_different_cat_p2_puzzles() -> anyhow::Result<()> {
+        let mut sim = Simulator::new();
+        let ctx = &mut SpendContext::new();
+        let (sk, pk, puzzle_hash, coin) = sim.new_p2(2)?;
+
+        // This will just return the solution verbatim.
+        let custom_p2 = ctx.alloc(&1)?;
```
Comment (Yakuhito) [inline crates/chia-sdk-driver/src/primitives/cat.rs:469]: you can also get multiple sks via `test_secret_keys(2)`

Comment (Rigidity) [reply crates/chia-sdk-driver/src/primitives/cat.rs:469]: But I only need one

File: crates/chia-sdk-driver/src/primitives/did.rs [hunk 2]
```diff
@@ -176,43 +227,123 @@ where
...
+        assert_eq!(did.info.p2_puzzle_hash, puzzle_hash);
 
-        let (create_did, did) = Launcher::new(coin.coin_id(), 1).create_simple_did(ctx, pk)?;
+        Ok(())
+    }
 
-        // Make sure that bounds are relaxed enough to do this.
-        let metadata_ptr = ctx.alloc(&did.info.metadata)?;
-        let mut did = did.with_metadata(metadata_ptr);
+    #[rstest]
```
Comment (Yakuhito) [inline crates/chia-sdk-driver/src/primitives/did.rs:260]: retest- yay!

Comment (Rigidity) [reply crates/chia-sdk-driver/src/primitives/did.rs:260]: lol

File: crates/chia-sdk-driver/src/spend_with_conditions.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+use chia_sdk_types::Conditions;
+use clvm_utils::TreeHash;
+
+use crate::{DriverError, Spend, SpendContext};
+
+pub trait SpendWithConditions {
+    fn puzzle_hash(&self) -> TreeHash;
```
Comment (Yakuhito) [inline crates/chia-sdk-driver/src/spend_with_conditions.rs:7]: Why aren't primitives just `ToTreeHash`?

Comment (Rigidity) [reply crates/chia-sdk-driver/src/spend_with_conditions.rs:7]: Good point, I could just add a ToTreeHash bound wherever this is used for now

Comment (Rigidity) [reply crates/chia-sdk-driver/src/spend_with_conditions.rs:7]: Or ToClvm<Allocator> and calculate the tree hash manually, idk

**Pull request merged**